src/utils/coroutine.js

```diff
@@ -139,13 +81,21 @@ export function coroutine(iter) {
         continue;
       }
       timers = _timers;
-      const { value, done } = doThrow ? iter.throw(nextValue) : iter.next(nextValue);
+      const v = doThrow ? iter.throw(nextValue) : iter.next(nextValue);
+      const done = v.done;
+      let value = v.value;
+
       doThrow = false;
       timers = null;
       if (done) {
         return value;
       }
 
+      if (isCancelablePromise(value)) {
+        rollbacks.push(value.rollback);
+        value = value.promise;
```

Comment:
Slightly weird that technically this could be a non-promise and would fail the check below. Typescript protects us from that, and I guess we end up throwing anway so its not a big deal, but just slightly odd how that works out.

Reply:
Yeah. `isCancelablePromise` doesn't actually check the `promise` part, does it...